### PR TITLE
#2549 - PT Request a Change - Financial Information (Ministry Processing)

### PR DESCRIPTION
**Pull request by andrepestana-aot**
- Changed the parttime workflows to calculated appeal variables in parallel:
image
image

- Updated places where those variables should be used:
image

image

-  Added E2E tests

### CODE REVIEW

**Review by sh16011993 — APPROVED** [review]
Nice Work @andrepestana-aot  👍 LGTM. Thank You for the PR walkthrough.

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn [hunk 1]
```diff
@@ -828,29 +822,63 @@
...
+        </zeebe:ioMapping>
+      </bpmn:extensionElements>
+      <bpmn:incoming>Flow_15xhu93</bpmn:incoming>
+      <bpmn:outgoing>Flow_1tkgnk6</bpmn:outgoing>
+    </bpmn:scriptTask>
+    <bpmn:intermediateThrowEvent id="Event_0asamuq" name="Calculate Financial Information">
+      <bpmn:extensionElements>
+        <zeebe:ioMapping>
+          <zeebe:output source="=if (is defined(appealsStudentFinancialInformationAppealData) and appealsStudentFinancialInformationAppealData != null)&#10;  then&#10;  appealsStudentFinancialInformationAppealData.taxReturnIncome &#10;else&#10;  studentDataTaxReturnIncome" target="calculatedDataTaxReturnIncome" />
+          <zeebe:output source="=if (is defined(appealsStudentFinancialInformationAppealData) and appealsStudentFinancialInformationAppealData != null)&#10;  then appealsStudentFinancialInformationAppealData.daycareCosts12YearsOrOver &#10;else&#10;  studentDataDaycareCosts12YearsOrOver" target="calculatedDataDaycareCosts12YearsOrOver" />
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn:855]: Since we started using the `is defined` the [FEEL Null check](https://docs.camunda.io/docs/next/components/modeler/feel/language-guide/feel-boolean-expressions/#null-check) was probably improved and we no longer need the `is defined`.
Can you please check if the below code works when the variable is not present?

```
if (appealsStudentFinancialInformationAppealData != null) then
  appealsStudentFinancialInformationAppealData.taxReturnIncome 
else
  studentDataTaxReturnIncome
```

There is a better operator to execute a coalesce-like comparison ([get or else](if (appealsStudentFinancialInformationAppealData != null) then
  appealsStudentFinancialInformationAppealData.taxReturnIncome 
else
  studentDataTaxReturnIncome)) but apparently it has an open issue right now: https://github.com/camunda/camunda-modeler/issues/3983

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn [hunk 2]
```diff
@@ -828,29 +822,63 @@
...
+      <bpmn:extensionElements>
+        <zeebe:script expression="=// Consider only dependents who were born on or before the study end date.&#10;inputDataDependents[date(item.dateOfBirth) &#60;= date(offeringStudyEndDate)]" resultVariable="calculatedDataDependants" />
+        <zeebe:ioMapping>
+          <zeebe:input source="=//TODO: Student appeal variable needs to be considered here.&#10;if studentDataDependants = null &#10;then []&#10;else studentDataDependants" target="inputDataDependents" />
+        </zeebe:ioMapping>
+      </bpmn:extensionElements>
+      <bpmn:incoming>Flow_15xhu93</bpmn:incoming>
+      <bpmn:outgoing>Flow_1tkgnk6</bpmn:outgoing>
+    </bpmn:scriptTask>
+    <bpmn:intermediateThrowEvent id="Event_0asamuq" name="Calculate Financial Information">
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn:851]: From the conversation in the chat I was expecting the use of the `Script task` instead of `Intermediate Throw Event`. Is there a reason to not use it or can we change?

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn [hunk 3]
```diff
@@ -828,29 +822,63 @@
...
+          <zeebe:output source="=if (is defined(appealsStudentFinancialInformationAppealData) and appealsStudentFinancialInformationAppealData != null)&#10;  then appealsStudentFinancialInformationAppealData.daycareCosts12YearsOrOver &#10;else&#10;  studentDataDaycareCosts12YearsOrOver" target="calculatedDataDaycareCosts12YearsOrOver" />
+          <zeebe:output source="=if (is defined(appealsStudentFinancialInformationAppealData) and appealsStudentFinancialInformationAppealData != null)&#10;  then appealsStudentFinancialInformationAppealData.daycareCosts11YearsOrUnder &#10;else&#10;  studentDataDaycareCosts11YearsOrUnder" target="calculatedDataDaycareCosts11YearsOrUnder" />
+        </zeebe:ioMapping>
+      </bpmn:extensionElements>
+      <bpmn:incoming>Flow_1vb0nqd</bpmn:incoming>
+      <bpmn:outgoing>Flow_0n8rikf</bpmn:outgoing>
+    </bpmn:intermediateThrowEvent>
+    <bpmn:intermediateThrowEvent id="Event_1mkiqso" name="Define PD/PPD Status">
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn:862]: Converting this to a `Script task` and using a similar approach from `inputDataDependents` would make this logic simpler.
The reason for this request is because this PR intention is also to set the foundation for upcoming tasks and to be used as a reference, hence having those changes in place would be useful.

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn [hunk 4]
```diff
@@ -828,29 +822,63 @@
...
+    <bpmn:sequenceFlow id="Flow_037lfnr" sourceRef="Gateway_0wuxqll" targetRef="Event_1mkiqso" />
+    <bpmn:sequenceFlow id="Flow_1vb0nqd" sourceRef="Gateway_0wuxqll" targetRef="Event_0asamuq" />
+    <bpmn:sequenceFlow id="Flow_0n8rikf" sourceRef="Event_0asamuq" targetRef="Gateway_08rr1xe" />
+    <bpmn:sequenceFlow id="Flow_1tkgnk6" sourceRef="Activity_1vwijyc" targetRef="Gateway_0wuxqll" />
+    <bpmn:parallelGateway id="Gateway_0wuxqll">
+      <bpmn:incoming>Flow_1tkgnk6</bpmn:incoming>
+      <bpmn:outgoing>Flow_037lfnr</bpmn:outgoing>
+      <bpmn:outgoing>Flow_1vb0nqd</bpmn:outgoing>
+    </bpmn:parallelGateway>
+    <bpmn:scriptTask id="Activity_1vwijyc" name="Define Dependents">
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn:841]: My understanding from the previous PR that introduced this task was that it would be integrated into the "Appeal" process, which means that it would be in the same parallel processing from PD/PDD and Financial Information appeals.
Please see the comment from [@dheepak-aot ](https://github.com/bcgov/SIMS/pull/3075#discussion_r1557987556) in the previous PR.

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2021-2022.bpmn [hunk 5]
```diff
@@ -1,5 +1,5 @@
...
+<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:zeebe="http://camunda.org/schema/zeebe/1.0" xmlns:modeler="http://camunda.org/schema/modeler/1.0" xmlns:camunda="http://camunda.org/schema/1.0/bpmn" id="Definitions_0ihicdz" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.20.0" modeler:executionPlatform="Camunda Cloud" modeler:executionPlatformVersion="8.1.0" camunda:diagramRelationId="7a75e87a-6774-4e60-9a94-5dbbe3be143a">
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2021-2022.bpmn:2]: We should not downgrade the editor version. Please keep it as "5.21.0" or higher.

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2021-2022.bpmn
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2021-2022.bpmn:1]: This file seems different from the others is it because we are expecting to remove it?

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2022-2023.bpmn [hunk 6]
```diff
@@ -276,26 +289,17 @@
       <bpmn:extensionElements>
         <zeebe:ioMapping>
           <zeebe:output source="=if&#10;  partner1CRAReportedIncome = null&#10;then&#10;  partner1TotalIncome&#10;else&#10;  partner1CRAReportedIncome" target="calculatedDataPartner1TotalIncome" />
-          <zeebe:output source="=if&#10;  studentDataCRAReportedIncome = null&#10;then&#10;  studentDataTaxReturnIncome&#10;else&#10;  studentDataCRAReportedIncome" target="calculatedDataStudentTotalIncome" />
+          <zeebe:output source="=if&#10;  studentDataCRAReportedIncome = null&#10;then&#10;  calculatedDataTaxReturnIncome&#10;else&#10;  studentDataCRAReportedIncome" target="calculatedDataStudentTotalIncome" />
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2022-2023.bpmn:292]: Why was this changed?

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn:1]: Can we mark the appeal section as below and create the `Script taks` using the same name from the UI as below?
image

File: sources/packages/backend/workflow/test/2022-2023/parttime-assessment/appeals/parttime-assessment-student-financial-information.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        createFakeStudentDependentEligibleForChildcareCost(
+          DependentChildCareEligibility.Eligible0To11YearsOld,
+          assessmentConsolidatedData.offeringStudyStartDate,
+        ),
+        createFakeStudentDependentEligibleForChildcareCost(
+          DependentChildCareEligibility.Eligible12YearsAndOver,
+          assessmentConsolidatedData.offeringStudyStartDate,
+        ),
+      ];
+      assessmentConsolidatedData.appealsStudentFinancialInformationAppealData =
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/test/2022-2023/parttime-assessment/appeals/parttime-assessment-student-financial-information.e2e-spec.ts:29]: `appealsStudentFinancialInformationAppealData` can be typed in the same way that we also have `StudentDependent` type defined in the `assessmentConsolidatedData`.

File: sources/packages/backend/workflow/test/models/assessment.model.ts [hunk 8]
```diff
@@ -76,6 +76,7 @@ export interface AssessmentConsolidatedData extends JSONDoc {
   appealsStudentIncomeAppealData?: JSONDoc;
   appealsPartnerIncomeAppealData?: JSONDoc;
   appealsStudentDisabilityAppealData?: JSONDoc;
+  appealsStudentFinancialInformationAppealData?: JSONDoc;
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/test/models/assessment.model.ts:79]: `appealsStudentFinancialInformationAppealData` can be typed in the same way that we also have `StudentDependent` type defined here.

File: sources/packages/backend/workflow/src/workflow-definitions/assessment-gateway.bpmn [hunk 9]
```diff
@@ -264,7 +264,7 @@
...
...careQuantity)&#10;    then calculatedDataDependantChildInDaycareQuantity else null,&#10;    dependantInfantQuantity: if is defined(calculatedDataDependantInfantQuantity)&#10;    then calculatedDataDependantInfantQuantity else null,&#10;    dependantDeclaredOnTaxesQuantity: if is defined(calculatedDataDependantDeclaredOnTaxesQuantity)&#10;    then calculatedDataDependantDeclaredOnTaxesQuantity else null,&#10;    dependantPostSecondaryQuantity: if is defined(calculatedDataDependantPostSecondaryQuantity)&#10;    then calculatedDataDependantPostSecondaryQuantity else null,&#10;    partnerStudentStudyWeeks: if is defined(calculatedDataPartnerStudentStudyWeeks)&#10;    then calculatedDataPartnerStudentStudyWeeks else null,&#10;    taxReturnIncome: if is defined(calculatedDataTaxReturnIncome)&#10;    then calculatedDataTaxReturnIncome else null,&#10;    daycareCosts12YearsOrOver: if is defined(calculatedDataDaycareCosts12YearsOrOver)&#10;    then calculatedDataDaycareCosts12YearsOrOver else null,&#10;    daycareCosts11YearsOrUnder: if is defined(calculatedDataDaycareCosts11YearsOrUnder)&#10;    then calculatedDataDaycareCosts11YearsOrUnder else null&#10;  }&#10;}" target="workflowData" />
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/src/workflow-definitions/assessment-gateway.bpmn:267]: Was there a requirement to start saving these values?
- taxReturnIncome
- daycareCosts12YearsOrOver
- daycareCosts11YearsOrUnder

**Review by andrewsignori-aot — COMMENTED** [review]
Nice work, please take a look at the comments.

File: sources/packages/backend/workflow/src/workflow-definitions/assessment-gateway.bpmn (hunk 9 above)
Comment (dheepak-aot) [inline sources/packages/backend/workflow/src/workflow-definitions/assessment-gateway.bpmn:267]: What is the reason behind these new properties added here ??

Comment (andrewsignori-aot) [reply sources/packages/backend/workflow/src/workflow-definitions/assessment-gateway.bpmn:267]: I guess that this is the same question from here: https://github.com/bcgov/SIMS/pull/3106#discussion_r1560168239
2min apart😄

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn (hunk 2 above)
Comment (andrepestana-aot) [reply sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn:851]: Yes, my bad. I got this on my mind that I should use the throw event component but you're right. I'm changing that.

File: sources/packages/backend/workflow/src/workflow-definitions/assessment-gateway.bpmn (hunk 9 above)
Comment (andrepestana-aot) [reply sources/packages/backend/workflow/src/workflow-definitions/assessment-gateway.bpmn]: No, there wasn't. Removed.

Comment (andrepestana-aot) [reply sources/packages/backend/workflow/src/workflow-definitions/assessment-gateway.bpmn]: No, there wasn't. Removed.

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn
Comment (andrepestana-aot) [reply sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn:1]: Ok. Added.

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn (hunk 3 above)
Comment (andrepestana-aot) [reply sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn:862]: Replaced the intermediate throw event with script task and it's simpler.

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn (hunk 4 above)
Comment (andrepestana-aot) [reply sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn:841]: I did that before, then later talking to Dheepak I undertood that would be dealt in another ticket. I moved that to be parallel with the other appeals.

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2021-2022.bpmn (hunk 5 above)
Comment (andrepestana-aot) [reply sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2021-2022.bpmn:2]: Sorry, did that my mistake.

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2021-2022.bpmn
Comment (andrepestana-aot) [reply sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2021-2022.bpmn:1]: Are we still updating it as we're not supporting that program year?

Comment (andrewsignori-aot) [reply sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2021-2022.bpmn:1]: It should be updated till no longer in Git.

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn (hunk 1 above)
Comment (andrepestana-aot) [reply sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn:855]: The FEEL null check works. Thanks.

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn [hunk 10]
```diff
@@ -828,30 +819,76 @@
...
+          <zeebe:input source="=//TODO: Student appeal variable needs to be considered here.&#10;if studentDataDependants = null &#10;then []&#10;else studentDataDependants" target="inputDataDependents" />
+        </zeebe:ioMapping>
+      </bpmn:extensionElements>
+      <bpmn:incoming>Flow_13hzh8q</bpmn:incoming>
+      <bpmn:outgoing>Flow_0xi2yp3</bpmn:outgoing>
+    </bpmn:scriptTask>
+    <bpmn:scriptTask id="Activity_1r2wvtx" name="Define Financial Information">
+      <bpmn:extensionElements>
+        <zeebe:ioMapping>
+          <zeebe:input source="=if appealsStudentFinancialInformationAppealData = null and appealsStudentFinancialInformationAppealData.taxReturnIncome = null &#10;  then studentDataTaxReturnIncome &#10;  else appealsStudentFinancialInformationAppealData.taxReturnIncome" target="inputDataDataTaxReturnIncome" />
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn:866]: I believe that `Data` is duplicated in the variable name.

File: sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn [hunk 11]
```diff
@@ -1,5 +1,5 @@
...
+<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:zeebe="http://camunda.org/schema/zeebe/1.0" xmlns:modeler="http://camunda.org/schema/modeler/1.0" xmlns:camunda="http://camunda.org/schema/1.0/bpmn" id="Definitions_0ihicdz" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.22.0" modeler:executionPlatform="Camunda Cloud" modeler:executionPlatformVersion="8.4.0" camunda:diagramRelationId="7a75e87a-6774-4e60-9a94-5dbbe3be143a">
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/src/workflow-definitions/parttime-assessment-2024-2025.bpmn:2]: Thanks for the upgrade to 8.4.0 also, we have been missing it for a while ❤️

**Review by andrewsignori-aot — APPROVED** [review]
Great work and thanks for making the changes, looks good 👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrepestana-aot
